internal/renew_root.go

```diff
@@ -0,0 +1,77 @@
+package internal
+
+import (
+	"bytes"
+	"encoding/pem"
+	"fmt"
+	"io"
+)
+
+type RootRenewalState struct {
+	BaseState
+	EstServer string
+}
+
+type rootRenewalContext = stateContext[*RootRenewalState]
+type rootRenewalStep = stateStep[*RootRenewalState]
+
+// Not type RootRenewalHandler stateHandler[*RootRenewalState].
+// We want methods from a parent to be inherited, thus use struct composition.
+type RootRenewalHandler struct {
+	stateHandler[*RootRenewalState]
+}
+
+func NewRootRenewalHandler(app *App, stateFile, estServer string) *RootRenewalHandler {
+	state := &RootRenewalState{EstServer: estServer}
+	return &RootRenewalHandler{
+		stateHandler[*RootRenewalState]{
+			stateContext: newStateContext[*RootRenewalState](app, stateFile, state),
+			steps: []rootRenewalStep{
+				fetchRootStep{},
+			},
+		},
+	}
+}
+
+func (h *RootRenewalHandler) Update() error {
+	return h.execute("RootCaUpdateStarted", "RootCaUpdateCompleted", true)
+}
+
+type fetchRootStep struct{}
+
+func (s fetchRootStep) Name() string {
+	return "Fetch new root"
+}
+
+func (s fetchRootStep) Execute(h *rootRenewalContext) error {
+	caFile := h.app.sota.GetOrDie("import.tls_cacert_path")
+	url := h.State.EstServer + "/cacerts"
+	res, err := h.client.Get(url)
```

Comment:
I think you can use these helpers which will help with retry and backoff- https://github.com/foundriesio/fioconfig/blob/main/internal/http.go#L84

Reply:
I _guess_ you are right... I just didn't pay enough attention as to why those helpers are not used for the [cert rotation](https://github.com/foundriesio/fioconfig/blob/main/internal/rotate_est.go#L88).
Maybe, in that case a "plain retry" would be dangerous because of the _POST_ which is not applicable for a simple fetch of new CAs?

Reply:
it was probably just me forgetting to use my own helper functions